### PR TITLE
Mulliken-related improvements

### LINKED ISSUE #452: Encapsulate Mulliken partitioning API
**Describe the current state of a part of the code**
The current API for using Mulliken's partitioning is using a raw vector `ao_list`.

**Describe a change that would improve the code base**
- [x] We can introduce a class

```cpp
class MullikenPartitioning {
private:
    // A set of AO indices that are included in the Mulliken-partitioning of an AO basis.
    std::vector<size_t> m_indices;

public:
    const std::vector<size_t>& indices() const;

    SquareMatrix<double> partitionMatrix() const;  // Return the partition matrix P_A

    RTransformationMatrix<double> projectionMatrix() const;  // Return the projection matrix C^-1 P_A C
};
```

- [x] Instances of this class may be generated through the addition of the following API on `RSpinOrbitalBasis`:
```cpp
MullikenPartition mullikenPartition(selector) const;
```

where `selector` is a user-supplied function that returns `true` if a shell as its argument should be included in the partitioning or not.

- [x] Finally, we can use the `MullikenPartition` instance by combining it with an `RSQOneElectronOperator`:
```cpp
RSQOneElectronOperator partitioned(MullikenPartition) const;
```

- [x] Effectively implementing the calculation of the [Mulliken-partitioned operator's matrix elements](https://gqcg-res.github.io/knowdes/mulliken-partitioning.html) can be done through a one-index transformation, as discussed in [spinor connections, which are used in perturbation-dependent Fock spaces](https://gqcg-res.github.io/knowdes/perturbation-dependent-fock-spaces.html).

- [x] Add 'selector' tests.
- [x] Improve Mulliken population tests.

### LINKED ISSUE #746: Implement Mulliken partitioning for unrestricted types
We have to be able to Mulliken-partition unrestricted types in order to start research on spin-redistribution.

- [x] check `USpinOrbitalBasis::calculateAtomicSpinZ`;
- [x] check `USpinOrbitalBasis::calculateMullikenOperator`;

### PR DESCRIPTION
**Pull request by lelemmen**
This PR adds some general improvements related to Mulliken-partitioning.
- [x] closes #452;
- [x] closes #746;

Since this PR also implements:
- generalized Mulliken partitioning;
- the quantization of S_z in an unrestricted spin-orbital basis (see also #383),

I've added an extra estimate of 2 directly to this PR.

**Don't forget:**
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

**Comment by lelemmen** [comment]
@xdvriend I'm still waiting for your review on this one.

**Comment by xdvriend** [comment]
Ow sorry! Completely lost track of this PR. I'll get to it right away!

File: gqcp/include/Basis/SpinorBasis/RSpinOrbitalBasis.hpp [hunk 1]
```diff
@@ -225,25 +232,30 @@ class RSpinOrbitalBasis:
...
-     *  @note this method is only available for real matrix representations
+     *  Partition this set of restricted spin-orbitals according to the Mulliken partitioning scheme.
+     * 
+     *  @param selector             A function that returns true for basis functions that should be included the Mulliken partitioning.
+     * 
+     *  @return A `RMullikenPartitioning` for the AOs selected by the supplied selector function.
      */
-    template <typename S = ExpansionScalar, typename = IsReal<S>>
-    ScalarRSQOneElectronOperator<double> calculateMullikenOperator(const std::vector<size_t>& ao_list) const {
+    RMullikenPartitioning<ExpansionScalar> mullikenPartitioning(const std::function<bool(const BasisFunction&)>& selector) const {
```
Comment (xdvriend) [inline gqcp/include/Basis/SpinorBasis/RSpinOrbitalBasis.hpp:241]: Does the selector mean that we no longer have to specify the "Mulliken indices" in our calculations? That is an awesome improvement!

**Review by xdvriend — APPROVED** [review]
Everything looks great to me! Again my apologies that I lost track of this PR and you had to wait for your review.

File: gqcp/include/Basis/SpinorBasis/RSpinOrbitalBasis.hpp (hunk 1 above)
Comment (lelemmen) [reply gqcp/include/Basis/SpinorBasis/RSpinOrbitalBasis.hpp:241]: Yes, you can check an example in the tests! :)

**Pull request merged**